boot/setefibootvars.go

```diff
@@ -0,0 +1,196 @@
+// path corresponding to the given asset path, along with the given description
+// and optional data.
+func ConstructLoadOption(description string, assetPath string, optionalData []byte) ([]byte, error) {
+	devicePath, err := linuxFilePathToDevicePath(assetPath, linux.ShortFormPathHD)
+	if err != nil {
+		return nil, err
+	}
+	loadOption := &efi.LoadOption{
+		Attributes:   efi.LoadOptionActive | efi.LoadOptionCategoryBoot,
+		Description:  description,
+		FilePath:     devicePath,
+		OptionalData: optionalData,
+	}
+	loadOptionSerialized, err := loadOption.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	return loadOptionSerialized, nil
+}
+
+// Searches existing Boot#### variables for one which matches the given data.
+//
+// If there is a match, returns the boot number of the existing variable.
+// Otherwise, finds the first available boot number and returns it, along with
+// ErrNoMatchingVariable, indicating that a new boot option with that boot
+// number should be written. If a different error occurs, returns that error,
+// and the returned boot number should be ignored.
+func findMatchingBootOption(optionData []byte) (uint16, error) {
+	variables, err := efiListVariables()
+	if err != nil {
+		return 0, err
+	}
+	bootVarRegexp, err := regexp.Compile("^Boot[0-9A-F]{4}$")
+	if err != nil {
+		// error should never occur
+		return 0, err
+	}
+	usedBootNums := make(map[uint64]bool)
+	for _, varDesc := range variables {
+		varName := varDesc.Name
+		varGUID := varDesc.GUID
+		if !bootVarRegexp.MatchString(varName) {
+			// Not a Boot#### variable, so skip it
+			continue
+		}
+		if varGUID != efi.GlobalVariable {
+			// Not an EFI global variable, so skip it
+			continue
+		}
+		varNumber, err := strconv.ParseUint(varName[4:], 16, 16)
+		if err != nil {
+			// Should not occur, since variable matched bootVarRegexp
+			return 0, err
+		}
+		// Since we never overwrite an existing variable, we can ignore
+		// variable attributes when reading the variable
+		varData, _, err := efiReadVariable(varName, varGUID)
+		if err != nil {
+			return 0, err
+		}
+		if reflect.DeepEqual(optionData, varData) {
+			// existing variable already identical, use it
+			return uint16(varNumber), nil
+		}
+		usedBootNums[varNumber] = true
+	}
+	for bootNum := uint64(0); bootNum <= 0xFFFF; bootNum++ {
+		if usedBootNums[bootNum] == false {
+			return uint16(bootNum), ErrNoMatchingVariable
+		}
+	}
+	return 0, ErrAllBootNumsUsed
+}
+
+// Ensures that a Boot#### variable contains the given EFI load option.
+//
+// It may be the case than an existing boot variable already contains the
+// given load option, in which case that boot variable is reused. Otherwise,
+// finds the first unused boot variable number and uses it. Writes the load
+// option to that variable, and returns the number of the variable that was
+// used.
+func SetEfiBootOptionVariable(loadOptionData []byte) (uint16, error) {
```

Comment:
This could be useful elsewhere (more likely than `ConstructLoadOption`, at least), but I'll export it as well. Do you think `SetEfiBootNumberVariable` should be exported if `SetEfiBootOrderVariable` is not? Perhaps only `SetEfiBootVariables` should be be exported, since if callers are independently able to read variable contents enough to know what to set as the boot order, they can probably just set the boot order in the first place, or they can formulate a load option and then call `SetEfiBootVariables`, which will not overwrite the . Perhaps we should still allow this functionality though. I'll keep `SetEfiBootOrder` and `SetEfiBootVariables` exported, and unexport the others.

Reply:
Usually we only export if functions need to be used in other packages. In this case we only really need `SetEfiBootVariables` - if in the end we really need to use something else we can just export as needed. Also, I see `SetEfiBootOptionVariable` and `SetEfiBootOrderVariable` to sort of come together, so they should be both exported or not. And it looks like they are not used outside of the package.

Reply:
That makes sense. I'll unexport `SetEfiBootOrderVariable` as well. Thanks!